Snowflake.Data/Core/SFResultSetArrow.cs

```diff
@@ -0,0 +1,176 @@
+        }
+
+        internal override async Task<bool> NextAsync()
+        {
+            if (isClosed)
+            {
+                throw new SnowflakeDbException(SFError.DATA_READER_ALREADY_CLOSED);
+            }
+
+            _currentChunkRowIdx++;
+            if (_currentChunkRowIdx < _currentChunkRowCount)
+            {
+                return true;
+            }
+
+            if (_chunkDownloader != null)
+            {
+                throw new SnowflakeDbException(SFError.UNSUPPORTED_FEATURE);
+            }
+
+            return false;
+        }
+
+        internal override bool Next()
+        {
+            if (isClosed)
+            {
+                throw new SnowflakeDbException(SFError.DATA_READER_ALREADY_CLOSED);
+            }
+
+            _currentChunkRowIdx++;
+            if (_currentChunkRowIdx < _currentChunkRowCount)
+            {
+                return true;
+            }
+
+            if (_chunkDownloader != null)
+            {
+                throw new SnowflakeDbException(SFError.UNSUPPORTED_FEATURE);
+            }
+
+            return false;
+        }
+
+        internal override bool NextResult()
+        {
+            return false;
+        }
+
+        internal override async Task<bool> NextResultAsync(CancellationToken cancellationToken)
+        {
+            return await Task.FromResult(false);
+        }
+
+        internal override bool HasRows()
+        {
+            if (isClosed)
+            {
+                return false;
+            }
+
+            return _currentChunkRowCount > 0 || _totalChunkCount > 0;
+        }
+
+        /// <summary>
+        /// Move cursor back one row.
+        /// </summary>
+        /// <returns>True if it works, false otherwise.</returns>
+        internal override bool Rewind()
+        {
+            if (isClosed)
+            {
+                throw new SnowflakeDbException(SFError.DATA_READER_ALREADY_CLOSED);
+            }
+
+            if (_currentChunkRowIdx >= 0)
+            {
+                _currentChunkRowIdx--;
+                if (_currentChunkRowIdx >= _currentChunkRowCount)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        internal override UTF8Buffer getObjectInternal(int columnIndex)
+        {
+            if (isClosed)
+            {
+                throw new SnowflakeDbException(SFError.DATA_READER_ALREADY_CLOSED);
+            }
+
+            if (columnIndex < 0 || columnIndex >= columnCount)
+            {
+                throw new SnowflakeDbException(SFError.COLUMN_INDEX_OUT_OF_BOUND, columnIndex);
+            }
+
+            return _currentChunk.ExtractCell(_currentChunkRowIdx, columnIndex);
+        }
+
+        private void updateSessionStatus(QueryExecResponseData responseData)
+        {
+            SFSession session = this.sfStatement.SfSession;
+            session.database = responseData.finalDatabaseName;
```

Comment:
could we call (new) method `session.update(responseData.finalDatabaseName, responseData.finalSchemaName)` or enrich `UpdateSessionParameterMap` with those parameters to make the code more object oriented?